src/Symfony/Component/HttpClient/HttpClientTrait.php

```diff
@@ -473,28 +473,27 @@ private static function mergeQueryString(?string $queryString, array $queryArray
         $query = [];
 
         if (null !== $queryString) {
-            foreach (explode('&', $queryString) as $v) {
-                if ('' !== $v) {
-                    $k = urldecode(explode('=', $v, 2)[0]);
-                    $query[$k] = (isset($query[$k]) ? $query[$k].'&' : '').$v;
-                }
-            }
+            parse_str($queryString, $query);
         }
 
-        foreach ($queryArray as $k => $v) {
-            if (is_scalar($v)) {
-                $queryArray[$k] = rawurlencode($k).'='.rawurlencode($v);
-            } elseif (null === $v) {
-                unset($queryArray[$k]);
-
-                if ($replace) {
-                    unset($query[$k]);
+        $iterator = new \RecursiveIteratorIterator(new \RecursiveArrayIterator($queryArray), \RecursiveIteratorIterator::SELF_FIRST);
```

Comment:
in order to merge things correctly, I'd suggest a different algorithme:
we should use `http_build_query` *before* merging with the query string
then we should parse the output using similar logic as L476